providers/mlx5/verbs.c

```diff
@@ -2933,32 +2933,158 @@ int mlx5_destroy_wq(struct ibv_wq *wq)
 	return 0;
 }
 
+static void free_flow_counters_descriptions(struct mlx5_ib_create_flow *cmd)
+{
+	int i;
+
+	for (i = 0; i < cmd->ncounters_data; i++)
+		free((void *)(uintptr_t)cmd->data[i].counters_data);
+}
+
+static int get_flow_mcounters(struct mlx5_flow *mflow,
+			      struct ibv_flow_attr *flow_attr,
+			      struct mlx5_counters **mcounters,
+			      uint32_t *data_size)
+{
+	struct ibv_flow_spec *ib_spec;
+	uint32_t ncounters_used = 0;
+	int i;
+
+	ib_spec = (struct ibv_flow_spec *)(flow_attr + 1);
+	for (i = 0; i < flow_attr->num_of_specs; i++, ib_spec = (void *)ib_spec + ib_spec->hdr.size) {
+		if (ib_spec->hdr.type != IBV_FLOW_SPEC_ACTION_COUNT)
+			continue;
+
+		/* currently support only one counters data */
+		if (ncounters_used > 0)
+			return EINVAL;
+
+		*mcounters  = to_mcounters(ib_spec->flow_count.counters);
+		ncounters_used++;
+	}
+
+	*data_size = ncounters_used * sizeof(struct mlx5_ib_flow_counters_data);
+	return 0;
+}
+
+static int allocate_flow_counters_descriptions(struct mlx5_counters *mcounters,
+					       struct mlx5_ib_create_flow *cmd)
+{
+	struct mlx5_ib_flow_counters_data *mcntrs_data;
+	uint32_t *cntrs_indexes, *cntrs_data;
+	struct mlx5_counter_node *cntr_node;
+	uint32_t ncounters;
+	int j = 0;
+
+	mcntrs_data = cmd->data;
+	ncounters = mcounters->ncounters;
+
+	/* mlx5_attach_counters_point_flow was never called */
+	if (!ncounters)
+		return EINVAL;
+
+	/* each counter has both index and description */
+	cntrs_data = calloc(ncounters * 2, sizeof(uint32_t));
```

Comment:
wrong us of calloc, *2 needs to be on the sizeof, see remarks on kernel patches.